src/main/scala/com/kinja/amqp/AmqpProducer.scala

```diff
@@ -16,24 +26,59 @@ import play.api.libs.json._
 
 import java.util.concurrent.TimeUnit
 
+import scala.concurrent.Future
+import scala.concurrent.duration._
+import scala.util.Try
+
 class AmqpProducer(
 	connection: ActorRef,
 	actorSystem: ActorSystem,
+	messageStore: MessageStore,
 	connectionTimeOut: Long,
+	askTimeout: Long,
 	logger: Slf4jLogger)(exchange: ExchangeParameters) {
 
 	private val channel: ActorRef = createChannel()
 
 	Amqp.waitForConnection(actorSystem, connection, channel).await(connectionTimeOut, TimeUnit.SECONDS)
 
-	def publish[A: Writes](routingKey: String, message: A): Unit = {
+	channel ! ConfirmSelect
+	channel ! AddConfirmListener(createConfirmListener)
+
+	def publish[A: Writes](routingKey: String, message: A, saveTimeMillis: Long = System.currentTimeMillis()): Future[Try[Unit]] = {
 		val json = Json.toJson(message)
 		val bytes = json.toString.getBytes(java.nio.charset.Charset.forName("UTF-8"))
 		val properties = new BasicProperties.Builder().deliveryMode(2).build()
-		channel ! Publish(exchange.name, routingKey, bytes, properties = Some(properties), mandatory = true, immediate = false)
+		implicit val timeout = Timeout(askTimeout.seconds)
+		import scala.concurrent.ExecutionContext.Implicits.global
+		channel ? Publish(exchange.name, routingKey, bytes, properties = Some(properties), mandatory = true, immediate = false) map { resp =>
+			resp match {
+				case ok: Ok => {
+					ok.result match {
+						case Some(MessageUniqueKey(deliveryTag, channelId)) =>
+							messageStore.saveMessage(Message(None, routingKey, json.toString, Some(channelId), Some(deliveryTag), saveTimeMillis))
+						case _ => messageStore.saveMessage(Message(None, routingKey, json.toString, None, None, saveTimeMillis))
+					}
+				}
+				case err: NotConnectedError => messageStore.saveMessage(Message(None, routingKey, json.toString, None, None, saveTimeMillis))
+				case _ => messageStore.saveMessage(Message(None, routingKey, json.toString, None, None, saveTimeMillis))
+			}
+		} recoverWith {
+			case _ => Future(messageStore.saveMessage(Message(None, routingKey, json.toString, None, None, saveTimeMillis)))
+		}
 	}
 
 	private def createChannel(): ActorRef = {
 		ConnectionOwner.createChildActor(connection, ChannelOwner.props(init = List(Record(DeclareExchange(exchange)))))
 	}
+
+	private def createConfirmListener: ActorRef = actorSystem.actorOf(Props(new Actor {
+		def receive = {
+			case HandleAck(deliveryTag, multiple, channelId) =>
+				messageStore.saveConfirmation(MessageConfirmation(None, channelId, deliveryTag, multiple))
```

Comment:
If we save the confirmation here now, than we would have 3 additional queries instead of one: save the confirmation now, and later the confirmation and the message. Now we would just delete the message, we don't have to save the confimation just to delete it later. So based on extra queries, we gain 66% less queries